ConcurrentModificationException in parallel build somewhere in p2 AbstractRepository

I just got the following in a parallel build.
- Tycho 4.0.11
- `-T 6`

I am not sure whether this this is something that needs to be synchronized in the Tycho layer (or retried, similar to #3698?), or if this is an actual bug in P2, since `AbstractRepositoryManager` already uses a `synchronized(repositoryLock) { ... }`, and it still fails.

```
[ERROR] ConcurrentModificationException
java.util.ConcurrentModificationException
    at java.util.LinkedHashMap$LinkedHashIterator.nextNode (LinkedHashMap.java:1023)
    at java.util.LinkedHashMap$LinkedEntryIterator.next (LinkedHashMap.java:1058)
    at java.util.LinkedHashMap$LinkedEntryIterator.next (LinkedHashMap.java:1055)
    at org.eclipse.equinox.internal.p2.core.helpers.OrderedProperties$UnmodifiableProperties.<init> (OrderedProperties.java:213)
    at org.eclipse.equinox.internal.p2.core.helpers.OrderedProperties.unmodifiableProperties (OrderedProperties.java:37)
    at org.eclipse.equinox.p2.repository.spi.AbstractRepository.getProperties (AbstractRepository.java:110)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.addRepository (AbstractRepositoryManager.java:132)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.basicGetRepository (AbstractRepositoryManager.java:197)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.loadRepository (AbstractRepositoryManager.java:678)
    at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.loadRepository (ArtifactRepositoryManager.java:131)
    at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.loadRepository (ArtifactRepositoryManager.java:125)
    at org.eclipse.tycho.p2maven.transport.P2MirrorDisablingArtifactRepositoryManager.loadRepository (P2MirrorDisablingArtifactRepositoryManager.java:94)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.lambda$loadRepository$1 (RemoteArtifactRepositoryManager.java:105)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.accessRepository (RemoteArtifactRepositoryManager.java:137)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.loadRepository (RemoteArtifactRepositoryManager.java:105)
    at org.eclipse.tycho.p2.repository.RepositoryArtifactProvider.loadRepository (RepositoryArtifactProvider.java:112)
    at org.eclipse.tycho.p2.repository.LazyArtifactRepository.getDelegate (LazyArtifactRepository.java:50)
    at org.eclipse.tycho.p2.repository.LazyArtifactRepository.contains (LazyArtifactRepository.java:87)
    at org.eclipse.tycho.p2.repository.RepositoryArtifactProvider.contains (RepositoryArtifactProvider.java:133)
    at org.eclipse.tycho.p2.repository.MirroringArtifactProvider.isFileAlreadyAvailable (MirroringArtifactProvider.java:439)
    at org.eclipse.tycho.p2.repository.CompositeArtifactProvider.lambda$isFileAlreadyAvailable$0 (CompositeArtifactProvider.java:173)
    at java.util.stream.MatchOps$1MatchSink.accept (MatchOps.java:90)
    at java.util.Spliterators$ArraySpliterator.tryAdvance (Spliterators.java:1034)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:129)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:527)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:513)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:499)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:230)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:196)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.anyMatch (ReferencePipeline.java:632)
    at org.eclipse.tycho.p2.repository.CompositeArtifactProvider.isFileAlreadyAvailable (CompositeArtifactProvider.java:173)
    at org.eclipse.tycho.p2resolver.TargetPlatformBaseImpl.isFileAlreadyAvailable (TargetPlatformBaseImpl.java:200)
    at org.eclipse.tycho.p2resolver.PreliminaryTargetPlatformImpl.isFileAlreadyAvailable (PreliminaryTargetPlatformImpl.java:39)
    at org.eclipse.tycho.core.resolver.DefaultP2ResolutionResult.lambda$new$1 (DefaultP2ResolutionResult.java:56)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:197)
    at java.util.HashMap$KeySpliterator.forEachRemaining (HashMap.java:1715)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:509)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:499)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:921)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect (ReferencePipeline.java:682)
    at org.eclipse.tycho.core.resolver.DefaultP2ResolutionResult.<init> (DefaultP2ResolutionResult.java:70)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.toResolutionResult (P2ResolverImpl.java:252)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.resolveDependencies (P2ResolverImpl.java:246)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.resolveTargetDependencies (P2ResolverImpl.java:124)
    at org.eclipse.tycho.p2resolver.P2DependencyResolver.doResolveDependencies (P2DependencyResolver.java:378)
    at org.eclipse.tycho.p2resolver.P2DependencyResolver.resolveDependencies (P2DependencyResolver.java:328)
    at org.eclipse.tycho.core.osgitools.AbstractTychoProject.lambda$getDependencyArtifacts$0 (AbstractTychoProject.java:92)
    at org.eclipse.tycho.core.osgitools.DefaultReactorProject$LazyValue.get (DefaultReactorProject.java:307)
    at org.eclipse.tycho.core.osgitools.DefaultReactorProject.computeContextValue (DefaultReactorProject.java:200)
    at org.eclipse.tycho.core.osgitools.AbstractTychoProject.getDependencyArtifacts (AbstractTychoProject.java:82)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:98)
    at org.eclipse.tycho.core.maven.TychoProjectExecutionListener.beforeProjectLifecycleExecution (TychoProjectExecutionListener.java:111)
    at org.apache.maven.lifecycle.internal.CompoundProjectExecutionListener.beforeProjectLifecycleExecution (CompoundProjectExecutionListener.java:42)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:103)
    at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:206)
    at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:71)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:572)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
[ERROR] com.sap.adt.sapgui.embedding: null: ConcurrentModificationException
java.lang.RuntimeException: com.sap.adt.sapgui.embedding: null
    at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:209)
    at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:71)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:572)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
Caused by: java.util.ConcurrentModificationException
    at java.util.LinkedHashMap$LinkedHashIterator.nextNode (LinkedHashMap.java:1023)
    at java.util.LinkedHashMap$LinkedEntryIterator.next (LinkedHashMap.java:1058)
    at java.util.LinkedHashMap$LinkedEntryIterator.next (LinkedHashMap.java:1055)
    at org.eclipse.equinox.internal.p2.core.helpers.OrderedProperties$UnmodifiableProperties.<init> (OrderedProperties.java:213)
    at org.eclipse.equinox.internal.p2.core.helpers.OrderedProperties.unmodifiableProperties (OrderedProperties.java:37)
    at org.eclipse.equinox.p2.repository.spi.AbstractRepository.getProperties (AbstractRepository.java:110)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.addRepository (AbstractRepositoryManager.java:132)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.basicGetRepository (AbstractRepositoryManager.java:197)
    at org.eclipse.equinox.internal.p2.repository.helpers.AbstractRepositoryManager.loadRepository (AbstractRepositoryManager.java:678)
    at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.loadRepository (ArtifactRepositoryManager.java:131)
    at org.eclipse.equinox.internal.p2.artifact.repository.ArtifactRepositoryManager.loadRepository (ArtifactRepositoryManager.java:125)
    at org.eclipse.tycho.p2maven.transport.P2MirrorDisablingArtifactRepositoryManager.loadRepository (P2MirrorDisablingArtifactRepositoryManager.java:94)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.lambda$loadRepository$1 (RemoteArtifactRepositoryManager.java:105)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.accessRepository (RemoteArtifactRepositoryManager.java:137)
    at org.eclipse.tycho.p2maven.transport.RemoteArtifactRepositoryManager.loadRepository (RemoteArtifactRepositoryManager.java:105)
    at org.eclipse.tycho.p2.repository.RepositoryArtifactProvider.loadRepository (RepositoryArtifactProvider.java:112)
    at org.eclipse.tycho.p2.repository.LazyArtifactRepository.getDelegate (LazyArtifactRepository.java:50)
    at org.eclipse.tycho.p2.repository.LazyArtifactRepository.contains (LazyArtifactRepository.java:87)
    at org.eclipse.tycho.p2.repository.RepositoryArtifactProvider.contains (RepositoryArtifactProvider.java:133)
    at org.eclipse.tycho.p2.repository.MirroringArtifactProvider.isFileAlreadyAvailable (MirroringArtifactProvider.java:439)
    at org.eclipse.tycho.p2.repository.CompositeArtifactProvider.lambda$isFileAlreadyAvailable$0 (CompositeArtifactProvider.java:173)
    at java.util.stream.MatchOps$1MatchSink.accept (MatchOps.java:90)
    at java.util.Spliterators$ArraySpliterator.tryAdvance (Spliterators.java:1034)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:129)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:527)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:513)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:499)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:230)
    at java.util.stream.MatchOps$MatchOp.evaluateSequential (MatchOps.java:196)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.anyMatch (ReferencePipeline.java:632)
    at org.eclipse.tycho.p2.repository.CompositeArtifactProvider.isFileAlreadyAvailable (CompositeArtifactProvider.java:173)
    at org.eclipse.tycho.p2resolver.TargetPlatformBaseImpl.isFileAlreadyAvailable (TargetPlatformBaseImpl.java:200)
    at org.eclipse.tycho.p2resolver.PreliminaryTargetPlatformImpl.isFileAlreadyAvailable (PreliminaryTargetPlatformImpl.java:39)
    at org.eclipse.tycho.core.resolver.DefaultP2ResolutionResult.lambda$new$1 (DefaultP2ResolutionResult.java:56)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:197)
    at java.util.HashMap$KeySpliterator.forEachRemaining (HashMap.java:1715)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:509)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:499)
    at java.util.stream.ReduceOps$ReduceOp.evaluateSequential (ReduceOps.java:921)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.collect (ReferencePipeline.java:682)
    at org.eclipse.tycho.core.resolver.DefaultP2ResolutionResult.<init> (DefaultP2ResolutionResult.java:70)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.toResolutionResult (P2ResolverImpl.java:252)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.resolveDependencies (P2ResolverImpl.java:246)
    at org.eclipse.tycho.p2resolver.P2ResolverImpl.resolveTargetDependencies (P2ResolverImpl.java:124)
    at org.eclipse.tycho.p2resolver.P2DependencyResolver.doResolveDependencies (P2DependencyResolver.java:378)
    at org.eclipse.tycho.p2resolver.P2DependencyResolver.resolveDependencies (P2DependencyResolver.java:328)
    at org.eclipse.tycho.core.osgitools.AbstractTychoProject.lambda$getDependencyArtifacts$0 (AbstractTychoProject.java:92)
    at org.eclipse.tycho.core.osgitools.DefaultReactorProject$LazyValue.get (DefaultReactorProject.java:307)
    at org.eclipse.tycho.core.osgitools.DefaultReactorProject.computeContextValue (DefaultReactorProject.java:200)
    at org.eclipse.tycho.core.osgitools.AbstractTychoProject.getDependencyArtifacts (AbstractTychoProject.java:82)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:98)
    at org.eclipse.tycho.core.maven.TychoProjectExecutionListener.beforeProjectLifecycleExecution (TychoProjectExecutionListener.java:111)
    at org.apache.maven.lifecycle.internal.CompoundProjectExecutionListener.beforeProjectLifecycleExecution (CompoundProjectExecutionListener.java:42)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:103)
    at io.takari.maven.builder.smart.SmartBuilderImpl.buildProject (SmartBuilderImpl.java:206)
    at io.takari.maven.builder.smart.SmartBuilderImpl$ProjectBuildTask.run (SmartBuilderImpl.java:71)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:572)
    at java.util.concurrent.FutureTask.run (FutureTask.java:317)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:642)
    at java.lang.Thread.run (Thread.java:1583)
```

--- a/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/P2MirrorDisablingArtifactRepositoryManager.java
+++ b/p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/transport/P2MirrorDisablingArtifactRepositoryManager.java
@@ -52,8 +52,10 @@ class P2MirrorDisablingArtifactRepositoryManager implements IArtifactRepositoryM
 	private static void stripMirrorsURLProperty(AbstractRepository<?> repository, Logger logger) {
         try {
             Map<?, ?> properties = getRepositoryProperties(repository);
-            Object removedConfiguration = properties.remove(IRepository.PROP_MIRRORS_URL);
-
+            Object removedConfiguration;
+            synchronized (repository) {
+                removedConfiguration = properties.remove(IRepository.PROP_MIRRORS_URL);
+            }
             if (removedConfiguration != null && logger.isDebugEnabled()) {
                 logger.debug("Removed 'p2.mirrorsURL' property in repository " + repository.getLocation());
             }

